classes/check/tasklatencycheck.php

```diff
@@ -0,0 +1,139 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+/**
+ * Individual cron task latency monitoring check check.
+ *
+ * @package    tool_heartbeat
+ * @copyright  2022
+ * @author     Peter Burnett <peterburnett@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ *
+ * This can be run either as a web api, or on the CLI. When run on the
+ * CLI it conforms to the Nagios plugin standard.
+ *
+ * See also:
+ *  - http://nagios.sourceforge.net/docs/3_0/pluginapi.html
+ *  - https://nagios-plugins.org/doc/guidelines.html#PLUGOUTPUT
+ */
+
+namespace tool_heartbeat\check;
+use core\check\check;
+use core\check\result;
+
+defined('MOODLE_INTERNAL') || die();
+require_once("$CFG->libdir/cronlib.php");
+
+class tasklatencycheck extends check {
+
+    /**
+     * Get the status of the monitored tasks.
+     *
+     * @return result
+     */
+    public function get_result(): result {
+        global $CFG, $DB;
+
+        $taskconfig = get_config('tool_heartbeat', 'tasklatencymonitoring');
+        $status = result::OK;
+        $messages = '';
+
+        $lockstats = isset($CFG->lock_factory) && $CFG->lock_factory === '\\tool_lockstats\\proxy_lock_factory';
+
+        $tasks = explode(PHP_EOL, $taskconfig);
+        foreach ($tasks as $taskconfig) {
+            list($taskclass, $runtime, $startdelay, $completiondelay) = explode(',', str_replace(' ', '', $taskconfig));
+            $task = \core\task\manager::get_scheduled_task($taskclass);
+
+            // Input validation.
+            $valid = $task !== false;
+
+            // Cast to int, will force non-int strings to 0, so we only need to care about negative time as invalid.
+            $valid &= ((int) $runtime >= 0);
+            $valid &= ((int) $startdelay >= 0);
+            $valid &= ((int) $completiondelay >= 0);
+
+            if (!$valid) {
+                return new result(result::ERROR, get_string('taskconfigbad', 'tool_heartbeat', $taskclass));
+            }
+
+            // Lets see if the task is currently running. We can only do this via lockstats.
+            if ($lockstats) {
+                $locked = $DB->record_exists('tool_lockstats_locks', ['resourcekey' => $taskclass]);
+            } else {
+                $locked = false;
+            }
+
+            // Now we can start checking all of the data against the task itself.
+            // First check for start time drift.
+            $starttime = $task->get_next_run_time();
+            if ($startdelay != 0 && ($starttime <= (time() - $startdelay * MINSECS))) {
+                // Lets see if the task is currently running. We can only do this via lockstats.
+                if (!$locked) {
```

Comment:
Perhaps this can be moved into the outer condition (since there is no else statement), and `$locked` can come first since it's less computationally expense to check. Also it seems like the description was copied from the block above, which doesn't seem correct here.